### PR TITLE
[Doc] Theme revamp: styling changes

### LINKED ISSUE #42052: [Docs] Ray docs theme revamp: Styling changes
### Description

1. Change the theme of the code block on light and dark mode to stata-dark - https://pygments.org/styles/
We should use the same theme for the code block on the landing page. 
2. Increase the spacing between left nav elements (maybe change padding to 0.4em?)
3. Hover styles: 
Hover color for light mode: #09326C
Hover color for dark mode:  #CCE0FF

### Link

_No response_

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR modifies a few global docs styles and makes `index.html` use the same `pygments`-based code highlighting as elsewhere in the docs.

- Instead of inserting the raw html for `index.html` using sphinx, I made the index page use its own template. This allows us to pass in a `pygments` highlighting function to the HTML context, which produces the same code highlighting used elsewhere in the docs.
- Renamed `splash.css` -> `index.css`, `splash.js` -> `index.js`

## Related issue number

Closes https://github.com/ray-project/ray/issues/42052.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/conf.py [hunk 1]
```diff
@@ -409,9 +408,9 @@ def add_custom_assets(
     if pagename == "train/train":
         app.add_css_file("css/ray-train.css")
     elif pagename == "index":
-        # CSS for HTML part of index.html
-        app.add_css_file("css/splash.css")
-        app.add_js_file("js/splash.js")
+        app.add_css_file("css/index.css")
+        app.add_js_file("js/index.js")
+        return "index.html"
```
Comment (aslonnie) [inline doc/source/conf.py:413]: why is this returning `"index.html"` here?

File: doc/source/_static/css/index.css [hunk 2]
```diff
@@ -62,7 +62,7 @@
 
 .tab-pane pre {
   margin: 0;
-  padding: 0;
+  padding: 0.5em 1em;
```
Comment (aslonnie) [inline doc/source/_static/css/index.css:65]: why the left and right paddings are different? is that intended?

File: doc/source/conf.py (hunk 1 above)
Comment (peytondmurray) [reply doc/source/conf.py:413]: If you return a string from a function triggered by an `html-page-context` event, sphinx will use that template instead of the default `page.html` template for that page. [There's a note about this in the docstring for that function](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-events), but I can add more explanation if that would be helpful.

File: doc/source/_static/css/index.css (hunk 2 above)
Comment (peytondmurray) [reply doc/source/_static/css/index.css:65]: If two numbers are specified for the padding property, [the first number applies to top and bottom, the second number applies to left and right](https://developer.mozilla.org/en-US/docs/Web/CSS/padding#syntax). I just thought it looked a bit nicer this way, but can change it if another spacing would be better.

File: doc/source/_templates/index.html [hunk 3]
```diff
@@ -1,153 +1,153 @@
...
+        >
+        {{ pygments_highlight_python("""
 from typing import Dict
 import numpy as np
 
 import ray
 
 # Step 1: Create a Ray Dataset from in-memory Numpy arrays.
-ds = ray.data.from_numpy(np.asarray(["Complete this", "for me"]))
+ds = ray.data.from_numpy(np.asarray(['Complete this', 'for me']))
```
Comment (aslonnie) [inline doc/source/_templates/index.html:132]: why are we making these changes? seems unrelated to the change required here.

if it has to (or if you want to), could you split them into another PR?

File: doc/source/conf.py (hunk 1 above)
Comment (aslonnie) [reply doc/source/conf.py:413]: thanks for the explanation. could you move the index page handling up into its own section, and add a comment:

````
if pagename == "index":
  app.add_css_file("css/index.css")
  app.add_js_file("js/index.js")
  return "index.html"  # uses a special template

# The following uses the default page.html template
if pagename == "train/tarin":
  ...
````

File: doc/source/custom_directives.py [hunk 4]
```diff
@@ -50,11 +54,6 @@ def feedback_form_url(project, page):
     )
 
 
-def update_context(app, pagename, templatename, context, doctree):
```
Comment (aslonnie) [inline doc/source/custom_directives.py:53]: could you explain how is this change related? why the `update_context` function needs to be removed and inlined now?

File: doc/source/_static/css/index.css (hunk 2 above)
Comment (aslonnie) [reply doc/source/_static/css/index.css:65]: ah, ok. thanks.

File: doc/source/_templates/index.html (hunk 3 above)
Comment (peytondmurray) [reply doc/source/_templates/index.html:132]: These changes are being made automatically because we have a `.pre-commit-config.yml` that runs `prettier`, ensuring that we have consistent code style for the html/css/js that makes up the docs. Previously, we had no code formatting or linting for the docs.

Since we introduced the `pre-commit` config relatively recently, we haven't run it everywhere because we didn't want to touch all files at once. Eventually we will replace `format.sh` with `pre-commit`.

One of the side effects of this is that we are gradually introducing consistent code styling into the docs; any time we touch a file and `prettier` runs, we get changes like this. IMO this approach is less jarring to begin with because pre-commit hooks don't reformat the entire repository, but there's an argument to be made for just making one large commit with the pre-commit hooks run on all files at the same time.

File: doc/source/custom_directives.py (hunk 4 above)
Comment (peytondmurray) [reply doc/source/custom_directives.py:53]: > could you explain how is this change related?

Strictly speaking this isn't a necessary change, but right now we have two functions which run on `html-page-context`, both of which add things to the context. This is confusing for people who want to understand the docs build system, and since I was touching the context anyway to add the pygments highlighter I thought it would be a good time to fix this antipattern.

Comment (aslonnie) [reply doc/source/custom_directives.py:53]: maybe separate this change into another PR?

File: doc/source/_templates/index.html (hunk 3 above)
Comment (aslonnie) [reply doc/source/_templates/index.html:132]: could you have a separate PR that:

- Checkin all changes required for passing `pre-commit run -a`
- Add `pre-commit run -a` as part of the premerge lint check"

we do not want PR's to have these unrelated changes that distracts the actual content.

File: doc/source/custom_directives.py (hunk 4 above)
Comment (peytondmurray) [reply doc/source/custom_directives.py:53]: Sure, will do.

File: doc/source/_templates/index.html (hunk 3 above)
Comment (peytondmurray) [reply doc/source/_templates/index.html:132]: I'd really like to have this as a premerge lint check, but we aren't enforcing it repo-wide yet.

So the reason it's this way is that we're in an awkward phase right now where we are poised to move to pre-commit, but still only actively enforcing `format-sh`. And sadly, `format.sh` does nothing for anything in the `docs/` directory :/

I'll update this PR to remove any pre-commit related changes, then disable pre-commit temporarily for this. Whenever ray-ci is ready to make the transition away from `format.sh`, we can run `pre-commit` to get `docs/` formatted.

Comment (peytondmurray) [reply doc/source/_templates/index.html:132]: @aslonnie Actually, looking more closely at this particular section: I need these quotes to be `'` and not `"` due to the fact that we are supplying python code to the pygments lexer from an HTML template. We actually need this to be single quoted for this to work - otherwise the templating engine gets confused.

Alternatively, it might work if we use triple `'''` to enclose the python code that gets passed to pygments, allowing us to keep `"` in the code snippet. Would that be preferable?

Comment (aslonnie) [reply doc/source/_templates/index.html:132]: yeah, probably keep it as `"`, I think ray repo's styling uses and enforced `"` for strings everywhere. the code here should be kept consistent.

**Review by aslonnie — COMMENTED** [review]
changes mostly look good.

please address the string quoting styling consistency issue.

File: doc/source/_templates/index.html (hunk 3 above)
Comment (peytondmurray) [reply doc/source/_templates/index.html]: Hmm, well some of the example python code here already uses `'`, so even on `master` we're not consistent with the rest of the python code base. In any case I'll make it all black-compliant.

**Comment by peytondmurray** [comment]
Thanks for the feedback, should be ready for merge pending approval from @simran-2797.

**Review by simran-2797 — APPROVED** [review]

**Review by aslonnie — APPROVED** [review]
